codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ShapeValueGenerator.java

```diff
@@ -79,66 +83,90 @@ public void writeShapeValueInline(GoWriter writer, Shape shape, Node params) {
-                    params.accept(new ShapeValueNodeVisitor(writer, this, shape));
-                });
+                listDeclShapeValue(writer, (CollectionShape) shape, params);
                 break;
 
             case MAP:
-                mapDeclShapeValue(writer, shape.asMapShape().get(), () -> {
-                    params.accept(new ShapeValueNodeVisitor(writer, this, shape));
-                });
+                mapDeclShapeValue(writer, shape.asMapShape().get(), params);
                 break;
 
             case UNION:
+                unionDeclShapeValue(writer, shape.asUnionShape().get(), params.expectObjectNode());
+                break;
+
             case DOCUMENT:
                 LOGGER.warning("Skipping " + shape.getType() + " shape type not supported, " + shape.getId());
                 writer.writeInline("nil");
                 break;
 
             default:
-
-                scalarWrapShapeValue(writer, shape, () -> {
-                    params.accept(new ShapeValueNodeVisitor(writer, this, shape));
-                });
+                writeScalarPointerInline(writer, shape, params);
         }
     }
 
     /**
      * Writes the declaration for a Go structure. Delegates to the runner for member fields within the structure.
      *
      * @param writer writer to write generated code with.
-     * @param shape  the structure shpae
-     * @param inner  lambda to populate struct member fields.
+     * @param shape  the structure shape
+     * @param params parameters to fill the generated shape declaration.
      */
-    protected void structDeclShapeValue(GoWriter writer, StructureShape shape, Runnable inner) {
+    protected void structDeclShapeValue(GoWriter writer, StructureShape shape, Node params) {
         Symbol symbol = symbolProvider.toSymbol(shape);
 
         writer.write("&$T{", symbol);
-        inner.run();
+        params.accept(new ShapeValueNodeVisitor(writer, this, shape));
         writer.writeInline("}");
     }
 
+    /**
+     * Writes the declaration for a Go union.
+     *
+     * @param writer writer to write generated code with.
+     * @param shape  the union shape.
+     * @param params the params.
+     */
+    protected void unionDeclShapeValue(GoWriter writer, UnionShape shape, ObjectNode params) {
+        Symbol symbol = symbolProvider.toSymbol(shape);
+        for (Map.Entry<StringNode, Node> entry : params.getMembers().entrySet()) {
+            Optional<MemberShape> member = shape.getMember(entry.getKey().toString());
+            if (member.isPresent()) {
+                Shape target = model.expectShape(member.get().getTarget());
+                Symbol memberSymbol = SymbolUtils.createValueSymbolBuilder(
+                        symbol.getName() + symbolProvider.toMemberName(member.get()),
+                        symbol.getNamespace()
+                ).build();
+
+                writer.writeInline("$T{Value: ", memberSymbol);
```

Comment:
Should this `$T` be a `&$T` since union's interface method? e.g. `is<UnionShapeName` should be a pointer receiver to prevent any confusion code bugs when users try to extract a union member from the interface. Without the pointer receiver for `is<UnionShapeName` user doesn't know if the SDK did pointer or not for the union member wrapper types, and they'll need to test both.

```go
switch tv := range somUnion.(type) {
    case *FooMemberBar:
         fmt.Printf("%T is pointer, %v", tv, tv.Value)

    case FooMemberBar:
         fmt.Printf("%T is value, %v", tv, tv.Value)
}
```